File: solutions/scala/src/test/scala/TimeUtilityTests.scala
import TimeUtilityTests.testCases
import org.scalamock.scalatest.MockFactory
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
import org.scalatest.prop.TableDrivenPropertyChecks
import org.scalatest.prop.Tables.Table

import java.time.LocalTime

class TimeUtilityTests
  extends AnyFlatSpec
    with Matchers
    with MockFactory
    with TableDrivenPropertyChecks {

  "TimeUtility" should "return a description at any time" in {
    forAll(testCases) {
      (hour, expectedDescription) =>
        val clockStub: Clock = stub[Clock]
        (clockStub.time _).when().returns(LocalTime.of(hour, 0))

        TimeUtility.timeOfDay(clockStub) should be(expectedDescription)
    }
  }
}

object TimeUtilityTests {
  private val testCases =
    Table(
      ("hour", "expectedDescription"),
      (0, "Night"),
      (0, "Night"),
      (4, "Night"),
      (6, "Morning"),
      (9, "Morning"),
      (12, "Afternoon"),
      (17, "Afternoon"),
      (18, "Evening"),
      (23, "Evening")
    )
}

File: solutions/java/src/test/java/unit/tests/intro/TimeUtilityTests.java
package unit.tests.intro;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.time.LocalTime;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class TimeUtilityTests {
    @ParameterizedTest
    @MethodSource("timeTestCases")
    void supportOperations(int hour, String expectedDescription) {
        var clockMock = mock(Clock.class);
        when(clockMock.now()).thenReturn(LocalTime.of(hour, 0));

        assertThat(
                new TimeUtility(clockMock)
                        .getTimeOfDay()
        ).isEqualTo(expectedDescription);
    }

    private static Stream<Arguments> timeTestCases() {
        return Stream.of(
                Arguments.of(0, "Night"),
                Arguments.of(0, "Night"),
                Arguments.of(4, "Night"),
                Arguments.of(6, "Morning"),
                Arguments.of(9, "Morning"),
                Arguments.of(12, "Afternoon"),
                Arguments.of(17, "Afternoon"),
                Arguments.of(18, "Evening"),
                Arguments.of(23, "Evening")
        );
    }
}